prover/prover.go

```diff
@@ -312,41 +319,39 @@ func (p *Prover) setApprovalAmount() error {
 		return err
 	}
 
-	receipt, err := rpc.WaitReceipt(context.Background(), p.rpc.L1, tx)
+	receipt, err := rpc.WaitReceipt(ctx, p.rpc.L1, tx)
 	if err != nil {
 		return err
 	}
 
-	log.Info("prover approved taikoL1 for taiko token", "txHash", receipt.TxHash.Hex())
+	log.Info("Approved TaikoL1 for taiko token", "txHash", receipt.TxHash.Hex())
 
-	allowance, err = p.rpc.TaikoToken.Allowance(
-		&bind.CallOpts{},
+	if allowance, err = p.rpc.TaikoToken.Allowance(
+		&bind.CallOpts{Context: ctx},
 		p.proverAddress,
 		p.cfg.TaikoL1Address,
-	)
-	if err != nil {
+	); err != nil {
 		return err
 	}
 
-	log.Info("new allowance for taikoL1 contract", "allowance", allowance.String())
+	log.Info("New allowance for TaikoL1 contract", "allowance", allowance.String())
 
 	return nil
 }
 
 // Start starts the main loop of the L2 block prover.
 func (p *Prover) Start() error {
-	if err := p.setApprovalAmount(); err != nil {
-		log.Crit("failed to set approval amount", "error", err)
-		return err
-	}
-
 	p.wg.Add(1)
 	p.initSubscription()
 
 	go func() {
 		if err := p.srv.Start(fmt.Sprintf(":%v", p.cfg.HTTPServerPort)); !errors.Is(err, http.ErrServerClosed) {
 			log.Crit("Failed to start http server", "error", err)
 		}
+
+		if err := p.setApprovalAmount(p.ctx); err != nil {
```

Comment:
i think `p.srv.Start` which calls `srv.echo.Start` is blocking, so this will only call `setApprovalAmount` after the server is shutdown via `.Close`, which only happens when the prover is shutdown or an error occurs. we should move this above `p.srv.Start` or into its own goroutine

Reply:
Moved this above `p.srv.Start`